homeassistant/components/denonavr/media_player.py

```diff
@@ -249,12 +249,33 @@ def __init__(
 
         self._telnet_was_healthy: bool | None = None
 
-    async def _telnet_callback(self, zone, event, parameter):
+        self._properties = [
+            k for k, v in self.__class__.__dict__.items() if isinstance(v, property)
+        ]
+        self._property_hash = hash(0)
+
+    async def _telnet_callback(self, zone, event, parameter) -> None:
         """Process a telnet command callback."""
+        # There are multiple checks implemented which reduce unnecessary updates of the ha state machine
         if zone != self._receiver.zone:
             return
-
-        self.async_write_ha_state()
+        # Some updates trigger multiple events like one for artist and one for title for one change
+        # We skip every event except the last one
+        if event == "NS" and not parameter.startswith("E4"):
+            return
+        if event == "TA" and not parameter.startwith("ANNAME"):
+            return
+        if event == "HD" and not parameter.startswith("ALBUM"):
+            return
+        # Some event types fire more events than there are changes.
+        # Thus, we check if there are changes before calling self.async_write_ha_state.
+        property_values = ""
+        for prop in self._properties:
+            property_values += str(getattr(self, prop, ""))
+        property_hash = hash(property_values)
+        if property_hash != self._property_hash:
+            self._property_hash = property_hash
+            self.async_write_ha_state()
```

Comment:
I'm not sure if its worth the extra complexity of checking properties here as `async_write_ha_state` already has this check https://github.com/home-assistant/core/blob/e937693d975f1afe1bb30bd9fd348e57bff25518/homeassistant/core.py#L1540

It would pay off if most states reject but be slower if most callbacks continue on

Reply:
```
2023-03-19 02:53:20.716 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 58
2023-03-19 02:53:42.264 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 585
2023-03-19 02:53:43.007 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 58
2023-03-19 02:53:43.491 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 56
2023-03-19 02:53:43.681 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 555
2023-03-19 02:53:44.623 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 55
2023-03-19 02:53:45.046 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 545
2023-03-19 02:53:53.530 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS CLV 50
2023-03-19 02:53:54.579 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS LEVC 50
2023-03-19 02:53:54.599 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS SMG MUS
2023-03-19 02:53:54.633 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main SI DVD
2023-03-19 02:53:54.781 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS ALSDSP OFF
2023-03-19 02:53:54.788 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS ALSSET ON
2023-03-19 02:53:54.813 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS ALSVAL 000
2023-03-19 02:53:54.896 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MS STEREO
2023-03-19 02:53:54.931 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS MDAX OFF
2023-03-19 02:53:55.036 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS HOSIPS OFF
2023-03-19 02:53:55.053 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS HOSIPM AUT
2023-03-19 02:53:55.107 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS MULTEQ:OFF
2023-03-19 02:53:55.129 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main PS DYNEQ OFF
2023-03-19 02:53:55.156 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS DYNVOL OFF
2023-03-19 02:53:55.207 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS REFLEV 0
2023-03-19 02:53:55.240 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS DRC OFF
2023-03-19 02:53:55.270 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS DELAY 000
2023-03-19 02:53:55.300 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS LFE 00
2023-03-19 02:53:55.354 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS BAS 50
2023-03-19 02:53:55.394 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS TRE 50
2023-03-19 02:53:55.465 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS TONE CTRL OFF
2023-03-19 02:53:55.499 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main PS HEQ OFF
2023-03-19 02:53:55.535 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS HOSSHP OFF
2023-03-19 02:53:55.652 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS INFAISSIG 02
2023-03-19 02:53:55.667 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have NOT changed, skipping state update: Main SS INFAISFSV 48K
2023-03-19 02:54:01.160 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 54
2023-03-19 02:54:01.763 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 535
2023-03-19 02:54:02.198 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 53
2023-03-19 02:54:02.644 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 525
2023-03-19 02:54:03.064 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 53
2023-03-19 02:54:03.460 WARNING (MainThread) [homeassistant.components.denonavr.media_player] Properties have changed, updating state: Main MV 535
```

Reply:
Could go either way on this one. The code looks fine so OK with me to leave it in

Reply:
I thought about this some more and realized there is a subtle bug here since this is a non-zero risk of a hash collision.  Its probably very rare, but we could end up not writing state when it actually has changed because of the collison.

If the hash matches we need to do an check on the actual properties to ensure they are actually the same and its not a hash collision

https://eng.lyft.com/hashing-and-equality-in-python-2ea8c738fb9d

Reply:
Maybe it makes sense to do the equality check up front instead... or remove the check and let the one in `async_fire` happen instead. My only source of data for hit/miss ratio is my testing on my device so I don't really have a good data set to tell if it needs to be kept or not.

Reply:
I don't think this check would create hash collisions. The properties of the class are of types `str`, `bool`, `Dict[str, str]` and `List[str]`.
Anyway, all I wanted to achieve is that HA is not flooded by things like the `PS` and `SS` visible in your logs. I did not know, that HA already checks this. With this new information, I prefer removing the check in order to remove complexity.